mono/metadata/handle.c

```diff
@@ -17,204 +19,161 @@
+	HandleChunk *cur = stack->bottom;
+	HandleChunk *last = stack->top;
 
-	*arena_stack = arena->prev;
+	if (!cur)
+		return;
 
-	for (chunk = arena->chunk; chunk; chunk = next) {
-		next = chunk->next;
-		memset (&chunk->handles [0], 0, sizeof (MonoHandleStorage) * HANDLES_PER_CHUNK);
-		if (next != NULL)
-			chunk_free (chunk);
+	while (cur) {
+		int i;
+		for (i = 0; i < cur->size; ++i)
+			func ((gpointer*)&cur->objects [i], gc_data);
+		if (cur == last)
+			break;
+		cur = cur->next;
 	}
 }
 
-static void
-arena_scan (gpointer addr, MonoGCMarkFunc mark_func, gpointer gc_data)
+void
+mono_stack_mark_record_size (MonoThreadInfo *info, HandleStackMark *stackmark, const char *func_name)
 {
-	MonoHandleArena *arena;
-	MonoHandleArenaChunk *chunk;
-	int i;
-
-	for (arena = *(MonoHandleArena**) addr; arena; arena = arena->prev) {
-		for (chunk = arena->chunk; chunk; chunk = chunk->next) {
-			for (i = 0; i < chunk->handles_size; ++i) {
-				if (chunk->handles [i].__private_obj != NULL)
-					mark_func (&chunk->handles [i].__private_obj, gc_data);
-			}
-		}
+	HandleStack *handles = (HandleStack *)info->handle_stack;
+	HandleChunk *cur = stackmark->chunk;
+	int size = -stackmark->size; //discard the starting point of the stack
+	while (cur) {
+		size += cur->size;
+		if (cur == handles->top)
+			break;
+		cur = cur->next;
 	}
-}
 
-static void
-initialize (void)
-{
-	arena_desc = mono_gc_make_root_descr_user (arena_scan);
+	if (size > THIS_IS_AN_OK_NUMBER_OF_HANDLES)
+		printf ("%s USED %d handles\n", func_name, size);
 }
 
+/* Checked build helpers */
 void
-mono_handle_arena_init (MonoHandleArena **arena_stack)
+mono_handle_verify (MonoRawHandle raw_handle)
 {
-	mono_lazy_initialize (&arena_status, initialize);
-	mono_gc_register_root ((char*) arena_stack, sizeof (MonoHandleArena*), arena_desc, MONO_ROOT_SOURCE_HANDLE, "runtime threads handle arena");
+	
 }
 
-void
-mono_handle_arena_cleanup (MonoHandleArena **arena_stack)
+
+/*
+ * Pop the stack until @stackmark and make @value the top value.
+ *
+ * @return the new handle for what @value points to 
+ */
+MonoRawHandle
+mono_stack_mark_pop_value (MonoThreadInfo *info, HandleStackMark *stackmark, MonoRawHandle value)
 {
-	mono_gc_deregister_root ((char*) arena_stack);
+	g_error ("impl me");
 }
 
-MonoHandleArena*
-mono_handle_arena_current (void)
+/* Temporary place for some of the handle enabled wrapper functions*/
+
+MonoStringHandle
+mono_string_new_handle (MonoDomain *domain, const char *data)
 {
-	return (MonoHandleArena*) mono_thread_info_current ()->handle_arena;
+	return MONO_HANDLE_NEW (MonoString, mono_string_new (domain, data));
 }
 
-MonoHandleArena**
-mono_handle_arena_current_addr (void)
+MonoArrayHandle
+mono_array_new_handle (MonoDomain *domain, MonoClass *eclass, uintptr_t n, MonoError *error)
 {
-	return (MonoHandleArena**) &mono_thread_info_current ()->handle_arena;
-}
+	return MONO_HANDLE_NEW (MonoArray, mono_array_new_checked (domain, eclass, n, error));
```

Comment:
This is going to return a non-`mono_null_value_handle` even in the case of an error.